src/flint/types/fmpz.pyx

```diff
@@ -683,11 +683,30 @@ cdef class fmpz(flint_scalar):
         return fmpz_is_probabprime(self.val)
 
     def is_perfect_power(self):
+        r"""
+        Return True if this integer is of the form `r^k`, False otherwise.
+ 
+            >>> fmpz(81).is_perfect_power()
+            True
+            >>> fmpz(1234).is_perfect_power()
+            False
+        """
         cdef int k
         cdef fmpz v = fmpz()
         k = fmpz_is_perfect_power(v.val, self.val)
         return k != 0
```

Comment:
We can have both methods but more useful than `is_perfect_power` would be a method that returns the power like SymPy's `perfect_power` function:
```python
In [1]: perfect_power(81)
Out[1]: (3, 4)

In [2]: perfect_power(80)
Out[2]: False
```
I'm not sure what is best to return if it is not a perfect power. Returning False is perhaps a bit strange...

Alternatives:

1. Raise an exception.
2. Return clearly invalid number like `(0, -1)`.
3. Return a tuple with a boolean like `(3, 4, True)` or `(0, -1, False)`.
4. Return None.

Maybe returning `False` is good as anything else...

Reply:
Or perhaps it can just be:
```python
>>> perfect_power(80)
(80, 1)
```
So `perfect_power(n)` always returns `(r, k)` such that `n = r^k` where `k` is either maximal or `1` if no maximal `k` exists:
```python
>>> perfect_power(1)
(1, 1)
```

Reply:
`is_perfect_power` was there before this PR, but when I saw it, I also thought it should best return `(r,k)` since that data is provided by the C function (in that case I would lean towards `None` if not a perfect power), so I'd be happy to change that.
Are you also suggesting it be renamed to `perfect_power`? That could make sense since one might expect a method called `is_something` to return a boolean. On the other hand, it seems the currently prevailing convention is to mirror the C library names, so maybe it should stay `is_perfect_power`?

Reply:
Sorry I missed the fact that the function was already there. I guess it is better not to change it then.

We could add a separate `.perfect_power()` method though to return `(r, k)`. The convention it seems in other places is that a perfect power should have `r,k>=2` although I guess that the `.perfect_power()` method should be consistent with `is_perfect_power()`.

Reply:
Regarding tests, `fmpz_is_perfect_power` appears well-tested already: https://github.com/flintlib/flint/blob/main/src/fmpz/test/t-is_perfect_power.c

Reply:
I think that would be good. The issue is though that `(r,k)` returned by `fmpz_is_perfect_power` is not guaranteed to have maximal `k`; the right way to proceed would be to implement that feature in flint's `fmpz_is_perfect_power` and then use it for python-flint to implement a method `.perfect_power()` you are suggesting.

Reply:
Ah, okay. I suppose we could just call `fmpz_is_perfect_power` in a loop until we've reduced down to a number that is not a perfect power.

Maybe returning a result with non-maximal `k` is useful. I would be more inclined to reduce as far as possible (i.e. find maximal `k`). That is what Sage's `perfect_power` apparently does:
https://doc.sagemath.org/html/en/reference/rings_standard/sage/rings/integer.html#sage.rings.integer.Integer.perfect_power

This seems like the sort of situation where what is best in python-flint might be a bit different from what Flint itself does. I'm not sure what the performance implications would be though of calling `fmpz_is_perfect_power` in a loop (with successively smaller arguments) compared to just calling it once.